designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,153 @@
+public sealed class NuGetInstalledPackage
+{
+    /// <summary>The package id.</summary>
+    public string Id { get; }
+
+    /// <summary>The project's requested package range for the package.</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be same as the installed package version.
+    /// If the project uses PackageReference, this is the version string in the project file, which may not match the resolved package version, and may not be single version string.
+    /// </remarks>
+    public string RequestedRange { get; }
+
+    /// <summary>The installed package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the same as requested range.
+    /// If the project uses PackageReference, this will be the resolved version.
+    /// </remarks>
+    public string Version { get; }
+
+    /// <summary>Path to the extracted package</summary>
+    /// <remarks>
+    /// When Visual Studio is connected to a Codespaces or Live Share environment, the path will be for the remote envionrment, not local.
+    /// This may be null if the package was not restored successfully.
+    /// </remarks>
+    public string InstallPath { get; }
+
+    // This class will hopefully use C# record types when that language feature becomes available, so make the constructor not-public, to prevent breaking change when records come out.
+    internal NuGetInstalledPackage(string id, string requestedRange, string version, string installPath);
+}
+
+/// <summary>Factory to create contract types</summary>
+/// <remarks>Trying to be forwards compatible with what C#9 records are going to be</remarks>
+public static class NuGetContractsFactory
+{
+    /// <summary>Create a <see cref="NuGetInstalledPackage"/></summary>
+    /// <param name="id">Package Id</param>
+    /// <param name="requestedRange">The requested range</param>
+    /// <param name="version">The installed version</param>
+    /// <param name="installPath">The package install path</param>
+    /// <returns><see cref="NuGetInstalledPackage"/></returns>
+    public static NuGetInstalledPackage CreateNuGetInstalledPackage(string id, string requestedRange, string version, string installPath);
+
+    /// <summary>Create a <see cref="InstalledPackageResultStatus"/></summary>
+    /// <param name="status"><see cref="InstalledPackageResultStatus"/></param>
+    /// <param name="packages">Read-only collection of <see cref="NuGetInstalledPackage"/></param>
+    /// <returns><see cref="InstalledPackagesResult"/></returns>
+    public static InstalledPackagesResult CreateGetInstalledPackagesResult(InstalledPackageResultStatus status, IReadOnlyCollection<NuGetInstalledPackage> packages);
+}
+```
+
+## Future Work
+
+* Other required APIs: (incomplete list)
+  * InstallPackageAsync
+  * UninstallPackageAsync
+
+## Considerations
+
+* Use `System.Guid` for project identifiers
```

Comment:
Looks like we're using guids :)

Reply:
Yes, talking to project system people, and I got the same feedback from the RPC contracts api reviewers group, other identifiers, like "unique id" can cause problems when projects are renamed. So multiple people recommended we use project guid.

Someone from the RPC contracts api reviewers group also recommended using `System.Guid`, rather than string. The two points I listed below were basically copy-paste from their advice :)